packages/doenetml-worker-rust/lib-doenetml-core/src/new_core/render/render_children.rs

```diff
@@ -0,0 +1,73 @@
+use crate::{
+    components::{ComponentEnum, ComponentNode},
+    new_core::graph_node::{GraphNode, StructureGraph},
+    ComponentIdx,
+};
+
+/// Object that lets you query for some information about your children.
+#[derive(Debug)]
+pub struct ChildQueryObject<'a> {
+    /// The graph node of `self`. All data about children that is returned will be in the context of this node.
+    self_graph_node: GraphNode,
+    components: &'a [ComponentEnum],
+    strings: &'a [String],
+    structure_graph: &'a StructureGraph,
+}
+
+impl<'a> ChildQueryObject<'a> {
+    pub fn new(
+        component_idx: ComponentIdx,
+        components: &'a [ComponentEnum],
+        strings: &'a [String],
+        structure_graph: &'a StructureGraph,
+    ) -> Self {
+        ChildQueryObject {
+            self_graph_node: GraphNode::Component(component_idx),
+            components,
+            strings,
+            structure_graph,
+        }
+    }
+
+    /// Get an iterator that will iterate through your children in order.
+    pub fn child_iter(&self) -> impl Iterator<Item = GraphNode> + 'a {
+        self.structure_graph.content_children(self.self_graph_node)
+    }
+
+    /// Get the type of the component specified by `node`. `node` must
+    /// be a `GraphNode::Component`.
+    pub fn get_component_type(&self, node: GraphNode) -> &str {
+        match node {
+            GraphNode::Component(idx) => self.components[idx].get_component_type(),
+            _ => panic!(
+                "Can only get the type of a GraphNode::Component, not {:?}",
+                node
+            ),
+        }
+    }
+
+    /// Get the string contents of the component specified by `node`. `node` must
+    /// be a `GraphNode::String`.
+    pub fn get_string(&self, node: GraphNode) -> &str {
+        match node {
+            GraphNode::String(idx) => &self.strings[idx],
+            _ => panic!(
+                "Can only get the string of a GraphNode::String, not {:?}",
+                node
+            ),
+        }
+    }
+}
+
+/// All components that can be rendered must implement this trait.
+/// It instructs Core how to render the children of a component.
+pub trait ComponentChildren<'a> {
```

Comment:
Could we provide implementations of the two simple cases via a derive macro, based on the `#[pass_though_children]` and `#[no_rendered_children]` attributes? Then, component authors would need to create their implementation of the trait only if they want more complicated behavior.

We could also further lower the bar for component authors by changing the return type to a vector, which I think will be easier to understand and implement than an iterator object. My guess is that we don't need the efficiency of an iterator for the return type, given that we will be doing something with the children whenever the function is called (though we do want the `child_query_object` argument to be an iterator so it could be lazy).

Reply:
Yeah. That can be changed. I didn't want to do a derive macro until we knew the final form, since rust analyzer is much more helpful at finding issues when no macros are involved.